librz/core/cio.c

```diff
@@ -1104,3 +894,202 @@ RZ_API RzCmdStatus rz_core_io_pcache_print(RzCore *core, RzIODesc *desc, RzCmdSt
+
+	char *str = strdup(s);
+	if (!str) {
+		return false;
+	}
+
+	int len = rz_str_unescape(str);
+	if (!rz_core_write_at(core, addr, (const ut8 *)str, len + 1)) {
+		RZ_LOG_ERROR("Could not write '%s' at %" PFMT64x "\n", s, addr);
+		free(str);
+		return false;
+	}
+	free(str);
+	return true;
+}
+
+/**
+ * \brief Transform a block of data at \p addr according to the operation \p op and the hexvalue \p hex
+ *
+ * \param core Reference to RzCore instance
+ * \param addr Where the block of data to modify starts
+ * \param op Operation to perform on the block of data
+ * \param hex Optional hex string that may be required by the specific operation
+ * \param hexlen Optional length of the \p hex string. Must be present if \p hex is specified.
+ * \param buflen Used to return the length of the returned buffer
+ * \return The transformed buffer
+ */
+RZ_API RZ_OWN ut8 *rz_core_transform_op(RzCore *core, ut64 addr, RzCoreWriteOp op, ut8 *hex, int hexlen, int *buflen) {
+	rz_return_val_if_fail(core, NULL);
+	rz_return_val_if_fail(!hex || hexlen >= 0, NULL);
+	rz_return_val_if_fail(buflen, NULL);
+
+	switch (op) {
+	case RZ_CORE_WRITE_OP_ADD:
+	case RZ_CORE_WRITE_OP_SUB:
+	case RZ_CORE_WRITE_OP_DIV:
+	case RZ_CORE_WRITE_OP_MUL:
+	case RZ_CORE_WRITE_OP_AND:
+	case RZ_CORE_WRITE_OP_OR:
+	case RZ_CORE_WRITE_OP_XOR:
+	case RZ_CORE_WRITE_OP_SHIFT_LEFT:
+	case RZ_CORE_WRITE_OP_SHIFT_RIGHT:
+		rz_return_val_if_fail(hex && hexlen >= 0, NULL);
+		break;
+	default:
+		break;
+	}
+
+	ut8 *buf = RZ_NEWS(ut8, core->blocksize);
+	if (!buf) {
+		return NULL;
+	}
+
+	int len = rz_io_nread_at(core->io, addr, buf, core->blocksize);
+	if (len < 0) {
+		free(buf);
+		return NULL;
+	}
+
+	for (int i = 0, j = 0; i < len; i++, j = (j + 1) % (hexlen ? hexlen : 1)) {
+		ut16 tmp16;
+		ut32 tmp32;
+		ut64 tmp64;
+		switch (op) {
+		case RZ_CORE_WRITE_OP_BYTESWAP2:
+			if (i + 1 < len) {
+				tmp16 = rz_read_le16(buf + i);
+				rz_write_be16(buf + i, tmp16);
+				i++;
+			}
+			break;
+		case RZ_CORE_WRITE_OP_BYTESWAP4:
+			if (i + 3 < len) {
+				tmp32 = rz_read_le32(buf + i);
+				rz_write_be32(buf + i, tmp32);
+				i += 3;
+			}
+			break;
+		case RZ_CORE_WRITE_OP_BYTESWAP8:
+			if (i + 7 < len) {
+				tmp64 = rz_read_le64(buf + i);
+				rz_write_be64(buf + i, tmp64);
+				i += 7;
+			}
+			break;
+		case RZ_CORE_WRITE_OP_ADD:
+			buf[i] += hex[j];
+			break;
+		case RZ_CORE_WRITE_OP_SUB:
+			buf[i] -= hex[j];
+			break;
+		case RZ_CORE_WRITE_OP_DIV:
+			buf[i] = hex[j] ? buf[i] / hex[j] : 0;
+			break;
+		case RZ_CORE_WRITE_OP_MUL:
+			buf[i] *= hex[j];
+			break;
+		case RZ_CORE_WRITE_OP_AND:
+			buf[i] &= hex[j];
+			break;
+		case RZ_CORE_WRITE_OP_OR:
+			buf[i] |= hex[j];
+			break;
+		case RZ_CORE_WRITE_OP_XOR:
+			buf[i] ^= hex[j];
+			break;
+		case RZ_CORE_WRITE_OP_SHIFT_LEFT:
+			buf[i] <<= hex[j];
+			break;
+		case RZ_CORE_WRITE_OP_SHIFT_RIGHT:
+			buf[i] >>= hex[j];
+			break;
+		default:
+			rz_warn_if_reached();
+			break;
```

Comment:
shouldn't we return or something?

Reply:
I think it's alright to not do anything. This is anyway a programming error that will be printed by `rz_warn_if_reached`. When this case is reached, the buf will remain the same, which I think is fine.